core/src/main/java/org/apache/iceberg/io/CachingFileIO.java

```diff
@@ -0,0 +1,404 @@
+      //  differentiate metadata files vs other files.
+      return new CachingInputFile(fileContentCache, inFile);
+    }
+    return inFile;
+  }
+
+  @Override
+  public OutputFile newOutputFile(String path) {
+    return wrappedIO.newOutputFile(path);
+  }
+
+  @Override
+  public void deleteFile(String path) {
+    wrappedIO.deleteFile(path);
+    if (fileContentCache != null) {
+      fileContentCache.cache.invalidate(path);
+    }
+  }
+
+  public ContentCache getCache() {
+    return fileContentCache;
+  }
+
+  public long cacheEstimatedSize() {
+    return fileContentCache == null ? 0 : fileContentCache.cache.estimatedSize();
+  }
+
+  public static class ContentCache {
+    private final long expireAfterAccessMs;
+    private final long maxTotalBytes;
+    private final long maxContentLength;
+    private final Cache<String, byte[]> cache;
+
+    private ContentCache(long expireAfterAccessMs, long maxTotalBytes, long maxContentLength) {
+      this.expireAfterAccessMs = expireAfterAccessMs;
+      this.maxTotalBytes = maxTotalBytes;
+      this.maxContentLength = maxContentLength;
+
+      Caffeine<Object, Object> builder = Caffeine.newBuilder();
+      if (expireAfterAccessMs > 0) {
+        builder = builder.expireAfterAccess(Duration.ofMillis(expireAfterAccessMs));
+      }
+
+      this.cache = builder.maximumWeight(maxTotalBytes)
+          .weigher((Weigher<String, byte[]>) (key, value) -> value.length)
+          .recordStats()
+          .build();
+    }
+
+    public long expireAfterAccess() {
+      return expireAfterAccessMs;
+    }
+
+    public long maxContentLength() {
+      return maxContentLength;
+    }
+
+    public long maxTotalBytes() {
+      return maxTotalBytes;
+    }
+
+    public Cache<String, byte[]> cache() {
+      return cache;
+    }
+
+    public CacheStats stats() {
+      return cache.stats();
+    }
+
+    @Override
+    public String toString() {
+      return getClass().getSimpleName() + '{' +
+          "expireAfterAccessMs=" + expireAfterAccessMs + ", " +
+          "maxContentLength=" + maxContentLength + ", " +
+          "maxTotalBytes=" + maxTotalBytes + ", " +
+          "cacheStats=" + cache.stats() + '}';
+    }
+  }
+
+  private static class CachingInputFile implements InputFile {
+    private final ContentCache contentCache;
+    private final InputFile wrappedInputFile;
+
+    private CachingInputFile(ContentCache cache, InputFile inFile) {
+      this.contentCache = cache;
+      this.wrappedInputFile = inFile;
+    }
+
+    @Override
+    public long getLength() {
+      byte[] buf = contentCache.cache.getIfPresent(location());
+      return (buf != null) ? buf.length : wrappedInputFile.getLength();
+    }
+
+    @Override
+    public SeekableInputStream newStream() {
+      try {
+        if (getLengthChecked() <= contentCache.maxContentLength()) {
+          return cachedStream();
```

Comment:
Looks like the intent is to implement a read-through cache. Can you clarify that in the Javadoc for this `FileIO`?

Reply:
Done.